### PR TITLE
Updates and Optimizes Wasserstein / FID metrics

### LINKED ISSUE #205: Optimize FID computation, and explore other options for `sqrtm`.

### PR DESCRIPTION
**Pull request by dallasfoster**
# Modulus Pull Request

## Description
Replaces the existing frechet inception distance (FID) metric, which used scipy's `sqrtm` function with a general purpose Wasserstein 2 metric written in pytorch. The new function is more stable than the previous implementation because it makes assumptions about positive definite matrices for better matrix square roots. It also adds overloading for W2/FID calculations for (batched) univariate and multivariate cases. 

Closes #205 

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

**Comment by dallasfoster** [comment]
Open to discussion about removing `calculate_fid_from_inception_stats` since it is purely a wrapper meant to not break diffusion workflows.

@mnabian There might be breaking changes since the old function required conversion from torch -> numpy for the FID calculation. I did my best to make sure that the relevant portions of the example are compliant.

File: modulus/metrics/general/wasserstein.py [hunk 1]
```diff
@@ -12,12 +12,113 @@
...
+        # Compute trace(sig0 + sig1 - 2*(sig0^1/2 * sig1 * sig0^1/2)^1/2) first
+
+        # Compute sig0^1/2 first using eigen decomposition.
+        vals0, vecs0 = torch.linalg.eigh(sigma0)
+        if torch.any(vals0 < 0.0):
+            warn(
+                "Warning! Some eigenvalues are less than zero and matrix is not positive definite."
+            )
+            vals0 = torch.nn.functional.relu(vals0)
+        sqrt_sig0 = torch.matmul(
```
Comment (akshaysubr) [inline modulus/metrics/general/wasserstein.py:67]: Maybe worth considering Newton-Schulz or other series expansion based methods here if performance is a consideration? See this paper for discussion and comparisons to truncated Taylor polynomials and Pade approximants: https://openreview.net/pdf?id=-AOEi-5VTU8

> To give a concrete example, when the batch size is 64, our MPA-Lya is 2.58X faster than NS iteration and 27.25X faster than SVD, while our MTP-Lya is 5.82X faster than the NS iteration and 61.32X faster than SVD.

File: modulus/metrics/general/wasserstein.py [hunk 2]
```diff
@@ -12,12 +12,113 @@
...
+            vecs0.transpose(-2, -1),
+        )
+
+        # Compute T = tr(sig0 + sig1 - 2* sqrtC)
+        if sigma_ndim > 2:
+            T = torch.vmap(torch.trace)(sigma0 + sigma1 - 2 * sqrtC)
+        else:
+            T = torch.trace(sigma0 + sigma1 - 2 * sqrtC)
+
+        return torch.norm((mu0 - mu1), p=2, dim=-1) ** 2 + T
```
Comment (akshaysubr) [inline modulus/metrics/general/wasserstein.py:93]: Why norm with order 2 and then square again instead of just square and sum?

File: modulus/metrics/general/wasserstein.py [hunk 3]
```diff
@@ -12,12 +12,113 @@
...
+            vals0 = torch.nn.functional.relu(vals0)
+        sqrt_sig0 = torch.matmul(
+            torch.matmul(vecs0, torch.diag_embed(torch.sqrt(vals0))),
+            vecs0.transpose(-2, -1),
+        )
+
+        # Compute C = (sig0^1/2 * sig1 * sig0^1/2)
+        C = torch.matmul(torch.matmul(sqrt_sig0, sigma1), sqrt_sig0)
+
+        # Compute Csqrt = sqrt( C )
```
Comment (akshaysubr) [inline modulus/metrics/general/wasserstein.py:75]: `sqrt(sigma^1/2 * sigma_ref * sigma^1/2) == sqrtm(np.dot(sigma, sigma_ref))`?

File: test/metrics/diffusion/test_fid.py [hunk 4]
```diff
@@ -13,17 +13,17 @@
...
 
 
-import numpy as np
 import pytest
+import torch
 
 from modulus.metrics.diffusion import calculate_fid_from_inception_stats
 
 
 def test_fid_calculation():
```
Comment (akshaysubr) [inline test/metrics/diffusion/test_fid.py:22]: Parameterize for CPU and GPU execution?

**Review by akshaysubr — APPROVED** [review]
Looks good to me, added a few minor comments

File: modulus/metrics/general/wasserstein.py (hunk 2 above)
Comment (dallasfoster) [reply modulus/metrics/general/wasserstein.py:93]: You are totally right, this is a remnant from prototyping that I had thought I fixed.

File: modulus/metrics/general/wasserstein.py (hunk 1 above)
Comment (dallasfoster) [reply modulus/metrics/general/wasserstein.py:67]: I like this idea, I'll run some benchmarks.

File: modulus/metrics/general/wasserstein.py (hunk 3 above)
Comment (dallasfoster) [reply modulus/metrics/general/wasserstein.py:75]: No, but the key here is the trace operator that wraps these operations.

```math
\mathrm{tr}{ \left( \left( \Sigma^{1/2} \Sigma_{ref} \Sigma^{1/2} \right)^{1/2} \right) } = \mathrm{tr}{ \left( \left( \Sigma \Sigma_{ref} \right)^{1/2} \right) }
```

File: test/metrics/test_metrics_general.py [hunk 5]
```diff
@@ -348,6 +341,70 @@ def test_crps(device, rtol: float = 1e-3, atol: float = 1e-3):
...
+        w_norm,
+        torch.zeros([1], dtype=torch.float32, device=device),
+        rtol=rtol,
+        atol=100 * atol,
+    )
+
+    x = torch.randn((100_000, 3), device=device, dtype=torch.float32)
+    y = torch.randn((100_000, 3), device=device, dtype=torch.float32)
+    mu_x = x.mean(dim=0)
+    sig_x = torch.cov(x.T)
```
Comment (NickGeneva) [inline test/metrics/test_metrics_general.py:378]: Why transpose here?

File: test/metrics/test_metrics_general.py [hunk 6]
```diff
@@ -348,6 +341,70 @@ def test_crps(device, rtol: float = 1e-3, atol: float = 1e-3):
...
+    mu_y = y.mean(dim=0)
+    sig_y = torch.cov(y.T)
+    w_norm = w.wasserstein_from_normal(mu_x, sig_x, mu_y, sig_y)
+    assert torch.allclose(
+        w_norm,
+        torch.zeros([1], dtype=torch.float32, device=device),
```
Comment (NickGeneva) [inline test/metrics/test_metrics_general.py:384]: Is it worth it to change one of these tests to offsetting a normal distribution mean by like 1 and seeing if the metric returns the correct distance vs all testing zero?

**Review by NickGeneva — APPROVED** [review]
Overall math looks good to me

**Comment by NickGeneva** [comment]
Update change log prior to merge as well. Thanks!

File: modulus/metrics/general/wasserstein.py (hunk 2 above)
Comment (NickGeneva) [reply modulus/metrics/general/wasserstein.py:93]: Its nice for following along with wikipedia :)

File: modulus/metrics/general/wasserstein.py (hunk 1 above)
Comment (NickGeneva) [reply modulus/metrics/general/wasserstein.py:67]: Curious how much performance of this calculation is important for the (diffusion?) models that use this. Any ideas?

Comment (dallasfoster) [reply modulus/metrics/general/wasserstein.py:67]: So I implemented Newton-Schulz and ran some tests with random `(100,100)` and `(1000,1000)` matrices. The timings for Newton-Schulz depend on the stopping criterion (some tolerance in successive iterates). The results are compiled from 1000 independent tests and are as follows (in ms):

```
100x100
Eigen Decomposition:
Mean: 3.8977         Std: 4.6846   Low: 2.5036         Max: 148.9305
Newton:
Mean: 2.6360         Std: 16.9361   Low: 1.1909         Max: 536.7705

1000x1000
Eigen Decomposition:
Mean: 25.462         Std: 5.679          Low: 22.706   Max: 201.514
Newton:
Mean: 13.857          Std: 14.799         Low: 12.264         Max: 481.065
```

Comment (dallasfoster) [reply modulus/metrics/general/wasserstein.py:67]: I will include them both in this PR. Maybe this is the start of a new `modulus.linalg`

Comment (dallasfoster) [reply modulus/metrics/general/wasserstein.py:67]: Upon further discussion, this will be postponed to a further PR.

File: test/metrics/test_metrics_general.py (hunk 6 above)
Comment (dallasfoster) [reply test/metrics/test_metrics_general.py:384]: I will implement a second test.
